File: 30DaysOfCode/day12_inheritance/inheritance.js
'use strict';

var _input = '';
var _index = 0;
process.stdin.on('data', (data) => { _input += data; });
process.stdin.on('end', () => {
    _input = _input.split(new RegExp('[ \n]+'));
    main();    
});
function read() { return _input[_index++]; }

class Person {
    constructor(firstName, lastName, identification) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.idNumber = identification;
    }
    
    printPerson() {
        console.log(
            "Name: " + this.lastName + ", " + this.firstName 
            + "\nID: " + this.idNumber
        )
    }
}

class Student extends Person {
    constructor(firstName, lastName, id, scores) {
        super(firstName, lastName, id);
        this.testScores = scores;
    }
    
    calculate() {
        let scoreAvg = 0;
        
        for (let i = 0; i < this.testScores.length; i++) {
            scoreAvg += this.testScores[i];
        }
        scoreAvg /= this.testScores.length;

        if (scoreAvg >= 90 && scoreAvg <= 100)
            return 'O';
        else if (scoreAvg >= 80 && scoreAvg < 90)
            return 'E';
        else if (scoreAvg >= 70 && scoreAvg < 80)
            return 'A';
        else if (scoreAvg >= 55 && scoreAvg < 70)
            return 'P';                
        else if (scoreAvg >= 40 && scoreAvg < 55)
            return 'D';                
        return 'T';
    }
}

function main() {
    let firstName = read()
    let lastName = read()
    let id = +read()
    let numScores = +read()
    let testScores = new Array(numScores)
    
    for (var i = 0; i < numScores; i++) {
        testScores[i] = +read()  
    }

    let s = new Student(firstName, lastName, id, testScores)
    s.printPerson()
    s.calculate()
    console.log('Grade: ' + s.calculate())
}

File: 30DaysOfCode/day12_inheritance/inheritance.rb
class Person
	def initialize(firstName, lastName, id)
		@firstName = firstName
		@lastName = lastName
		@id = id
	end
	def printPerson()
		print("Name: ",@lastName , ", " + @firstName ,"\nID: " , @id)
	end
end

class Student < Person
    def initialize(firstName, lastName, id, scores)
        super(firstName, lastName, id)
        @scores = scores
    end
    
    def calculate()
        scoreAvg = 0
        for i in (0..(@scores.length - 1))
           scoreAvg += @scores[i]
        end
        scoreAvg /= @scores.length
        
        if (scoreAvg >= 90 and scoreAvg <= 100)
            return 'O'
        elsif (scoreAvg >= 80 and scoreAvg < 90)
            return 'E'
        elsif (scoreAvg >= 70 and scoreAvg < 80)
            return 'A'
        elsif (scoreAvg >= 55 and scoreAvg < 70)
            return 'P'
        elsif (scoreAvg >= 40 and scoreAvg < 45)
            return 'D'
        end
        return 'T'
    end
end

input = gets.split()
firstName = input[0]
lastName = input[1]
id = input[2].to_i
numScores = gets.to_i
scores = gets.strip().split().map!(&:to_i)
s = Student.new(firstName, lastName, id, scores)
s.printPerson           
print("\nGrade: " + s.calculate)
